File: backend/app/schemas/analytics.py
from datetime import datetime

from pydantic import BaseModel


class MetricPoint(BaseModel):
    label: str
    value: int | float


class TimeSeriesPoint(BaseModel):
    month: str
    count: int


class BottleneckPoint(BaseModel):
    node_key: str
    node_title: str | None
    workflow_name: str | None
    average_wait_seconds: float
    sample_size: int


class ActivityItem(BaseModel):
    id: int
    actor_name: str | None
    action: str
    entity_type: str
    entity_id: str | None
    description: str
    created_at: datetime


class ApprovalSnapshot(BaseModel):
    id: int
    workflow_instance_id: int
    reference_number: str | None
    workflow_name: str | None
    requester_name: str | None
    node_title: str | None
    status: str
    created_at: datetime


class RequestSnapshot(BaseModel):
    id: int
    reference_number: str
    workflow_name: str | None
    requester_name: str | None
    requester_department_name: str | None
    current_stage_title: str | None
    status: str
    started_at: datetime
    updated_at: datetime


class AnalyticsDashboard(BaseModel):
    total_executions: int
    completed: int
    rejected: int
    running: int
    completion_rate: float
    average_processing_seconds: float
    requests_per_month: list[TimeSeriesPoint]
    executions_by_workflow: list[MetricPoint]
    executions_by_department: list[MetricPoint]
    approval_outcomes: list[MetricPoint]
    completed_vs_rejected: list[MetricPoint]
    bottlenecks: list[BottleneckPoint]
    recent_activity: list[ActivityItem]
    pending_approvals: list[ApprovalSnapshot]
    recent_requests: list[RequestSnapshot]


class WorkflowAnalytics(BaseModel):
    workflow_id: int
    workflow_name: str
    total_executions: int
    completed: int
    rejected: int
    running: int
    completion_rate: float
    average_processing_seconds: float
    status_breakdown: list[MetricPoint]
    bottlenecks: list[BottleneckPoint]

File: frontend/src/types/analytics.ts
export type MetricPoint = {
  label: string;
  value: number;
};

export type TimeSeriesPoint = {
  month: string;
  count: number;
};

export type BottleneckPoint = {
  node_key: string;
  node_title: string | null;
  workflow_name: string | null;
  average_wait_seconds: number;
  sample_size: number;
};

export type ActivityItem = {
  id: number;
  actor_name: string | null;
  action: string;
  entity_type: string;
  entity_id: string | null;
  description: string;
  created_at: string;
};

export type ApprovalSnapshot = {
  id: number;
  workflow_instance_id: number;
  reference_number: string | null;
  workflow_name: string | null;
  requester_name: string | null;
  node_title: string | null;
  status: string;
  created_at: string;
};

export type RequestSnapshot = {
  id: number;
  reference_number: string;
  workflow_name: string | null;
  requester_name: string | null;
  requester_department_name: string | null;
  current_stage_title: string | null;
  status: string;
  started_at: string;
  updated_at: string;
};

export type AnalyticsDashboard = {
  total_executions: number;
  completed: number;
  rejected: number;
  running: number;
  completion_rate: number;
  average_processing_seconds: number;
  requests_per_month: TimeSeriesPoint[];
  executions_by_workflow: MetricPoint[];
  executions_by_department: MetricPoint[];
  approval_outcomes: MetricPoint[];
  completed_vs_rejected: MetricPoint[];
  bottlenecks: BottleneckPoint[];
  recent_activity: ActivityItem[];
  pending_approvals: ApprovalSnapshot[];
  recent_requests: RequestSnapshot[];
};
